src/utils/permissions.ts

```diff
@@ -191,8 +112,134 @@ export async function hasChannelPermissions(
+}
+
+export async function calculateServerPermissions(
+  memberID: string,
+  guildID: string,
+  permissions: Permission[],
+): Promise<true | "GUILD_NOT_FOUND" | "MEMBER_NOT_FOUND" | Permission> {
+  const guild = await cacheHandlers.get("guilds", guildID);
+  if (!guild) return "GUILD_NOT_FOUND";
+
+  // Check if the bot is the owner of the guild, if it is, returns true
+  if (memberID === guild.ownerID) return true;
+
+  const member = (await cacheHandlers.get("members", botID))?.guilds.get(
+    guildID,
+  );
+  if (!member) return "MEMBER_NOT_FOUND";
+
+  const permissionBits = [...member.roles, guild.id]
+    .map((id) => guild.roles.get(id)!)
+    // Remove any edge case undefined
+    .filter((role) => role)
+    .reduce((bits, data) => {
+      bits |= BigInt(data.permissions);
+
+      return bits;
+    }, BigInt(0));
+
+  if (permissionBits & BigInt(Permissions.ADMINISTRATOR)) return true;
+
+  let missingPerm: any;
+
+  permissions.forEach((permission) => {
+    if (!(permissionBits & BigInt(Permissions[permission]))) {
+      missingPerm = permission;
+      return;
+    }
+  });
+
+  if (missingPerm) return missingPerm;
+
+  return true;
+}
+
+/** Checks if the member has this permission. If the member is an owner or has admin perms it will always be true. */
+export async function memberHasPermission(
+  memberID: string,
+  guildID: string,
+  permissions: Permission[],
+) {
+  return calculateServerPermissions(memberID, guildID, permissions);
+}
+
+export async function botHasPermission(
+  guildID: string,
+  permissions: Permission[],
+) {
+  const calculated = await calculateServerPermissions(
+    botID,
+    guildID,
+    permissions,
+  );
+  if (calculated === true) return true;
+
+  return false;
+}
+
+export async function botThrowOnMissingPermission(
+  guildID: string,
+  permissions: Permission[],
+) {
+  const calculated = await calculateServerPermissions(
+    botID,
+    guildID,
+    permissions,
+  );
+  if (calculated === true) return true;
+
+  throw new Error(Errors[`MISSING_${calculated}` as Errors]);
+}
+
+/** Checks if the bot has the permissions in a channel */
+export function botHasChannelPermissions(
+  channelID: string,
+  permissions: Permission[],
+) {
+  return hasChannelPermissions(channelID, botID, permissions);
+}
+
+/** Checks if a user has permissions in a channel. */
+export async function hasChannelPermissions(
+  channelID: string,
+  memberID: string,
+  permissions: Permission[],
+) {
+  const calculated = await calculateChannelPermissions(
+    channelID,
+    memberID,
+    permissions,
+  );
+  if (calculated === true) return true;
+
+  return false;
+}
+
+/** Checks if the bot has the permissions in a channel, if not error will thrown */
+export async function botThrowOnMissingChannelPermission(
+  channelID: string,
+  permissions: Permission[],
+) {
+  return await throwOnMissingChannelPermission(channelID, botID, permissions);
```

Comment:
returns automatically await
```suggestion
  return throwOnMissingChannelPermission(channelID, botID, permissions);
```